### ISSUE TITLE
Interceptor not working.

### ISSUE THREAD

**Issue by paulvanbladel** [dgrubelic/vue-authenticate#33]
Thanks for this excellent library.
I'm currently connecting vue-authenticate with IdentityServer (Oauth2 and I'm using the Authorization CODE flow) and I'm at the point where I receive a valid bearer token from the IdentityServer. The token is stored in local storage with key name: vue-authenticate.vueauth_token.

the problem I'm facing is that the token is not sent automatically on each request towards the API.
I'm using vue-resource, so did i understand correctly that no explicit configuration is necessary ?

This is my main.js, Am I doing something wrong?

```
import Vue from 'vue'
import Quasar from 'quasar'
import router from './router'
import vuelidate from 'vuelidate'
import VueResource from 'vue-resource'
import VueAuthenticate from 'vue-authenticate'
Vue.use(Quasar) // Install Quasar Framework
Vue.use(vuelidate)
Vue.use(VueResource)

Vue.use(VueAuthenticate, {
  baseUrl: 'http://localhost:8080',
  providers: {
    identSrv: {
      name: 'identSrv',
      url: 'Token/Exchange',
      authorizationEndpoint: 'http://localhost:5000/connect/authorize', 
      redirectUri: window.location.origin || window.location.protocol + '//' + window.location.host,
      scope: ['profile', 'openid', 'MyApi'],
      responseType: 'code',
      scopePrefix: '',
      scopeDelimiter: ' ',
      requiredUrlParams: ['scope', 'nonce'],
      optionalUrlParams: ['display', 'state'],
      state: function () {
        var val = ((Date.now() + Math.random()) * Math.random()).toString().replace('.', '')
        return encodeURIComponent(val)
      },
      display: 'popup',
      oauthType: '2.0',
      clientId: 'PanelButlerVueJs',

      nonce: function () {
        var val = ((Date.now() + Math.random()) * Math.random()).toString().replace('.', '')
        return encodeURIComponent(val)
      },
      popupOptions: { width: 452, height: 633 }
    }
  }
})

```

**Comment by vejja** [comment]
+1

**Comment by dgrubelic** [comment]
Please, no need to "+1". If you want to get updates on this topic, subscribe to the issue.

**Comment by dgrubelic** [comment]
@paulvanbladel have you tried to name this provider with generic name "oauth2", like this:

```javascript
...
providers: {
  oauth2: {
    name: 'oauth2',
    url: 'Token/Exchange',
    // etc...
  }
}
...
```

**Comment by paulvanbladel** [comment]
Hi
Thanks a lot for the reply.
I didn't use oauth naming yet but i will try that.
Note that the info from the config section is digested already since i get the token correctly in the browser storage. It goes wrong when the jwt needs to be attached to a new http req.

**Comment by dgrubelic** [comment]
What request library are you using on server side? And what does your response look like when your server is returning auth token?

**Comment by paulvanbladel** [comment]
I tried changing the provider id string "identSrv" to "oauth2" but that didn't make any difference.
I'm using the authorization code flow and I'm using server side .net core which does the exchange of the authorization code with an an access_token. Exactly this same code is used (and it works fine) for doing the same flow also with satellizer (in the context of an angular App) and with aurelia-auth (in the context of an aurelia app).
This is the payload returned from my backend:

{"token":"eyJhbGciOiJSUzI1NiIsImtpZCI6ImI5NTRjMmVlNjkzNGExYTJhNWY4NDY5Njk2NTc1N2Q0IiwidHlwIjoiSldUIn0.eyJuYmYiOjE0OTgyOTEwMjcsImV4cCI6MTQ5ODI5MTE0NywiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo1MDAwIiwiYXVkIjpbImh0dHA6Ly9sb2NhbGhvc3Q6NTAwMC9yZXNvdXJjZXMiLCJQYW5lbEJ1dGxlckFwaSJdLCJjbGllbnRfaWQiOiJQYW5lbEJ1dGxlclZ1ZUpzIiwic3ViIjoiZGJlZmYxYWQtMmUwNC00NjY5LTgwMzItODk1MTMzMDNmM2M3IiwiYXV0aF90aW1lIjoxNDk4MjkxMDI3LCJpZHAiOiJsb2NhbCIsImdpdmVuX25hbWUiOiJqb3NAam9zLmNvbSIsInJvbGUiOlsidXNlciIsImRhdGFFdmVudFJlY29yZHMudXNlciIsImRhdGFFdmVudFJlY29yZHMiLCJzZWN1cmVkRmlsZXMudXNlciIsInNlY3VyZWRGaWxlcyJdLCJlbWFpbCI6Impvc0Bqb3MuY29tIiwic2NvcGUiOlsib3BlbmlkIiwicHJvZmlsZSIsIlBhbmVsQnV0bGVyQXBpIl0sImFtciI6WyJwd2QiXX0.BevhBje12ujiBZOo8FEJbP43HyYucAIW0h_NYzL3VyXRfXXWOS_R36ECIWu9Vqd2iSSa9riWHnpDGwrvDlNWLuGOAm76cfYKeCOo8_o_-ZSWdytSmn91hZNil_J8weTJpcCMSxyBMcd5XD5p3_RDwUHGiv086yMTp39E33ngjm9gxpohGn9mdQTP-zxbxJ7iYOW94MVLi4N32H1zPHe26G4kG8p_72LUSrZO7JRbavRflEmVWUDWeoCEN0ZggNW8xyX036M1NqXPgPOx5ATH-92iRJ7y2jeMAepIpGjQAOt5AitjufBt6mM_KfoUmYwWmXIqKxP8MdITTs4gHOZNLA"}

**Comment by paulvanbladel** [comment]
this is how it looks in browser storage:
![oauth](https://user-images.githubusercontent.com/4886272/27506930-788fa628-58c4-11e7-8bd0-c725c5cef316.png)

**Comment by dgrubelic** [comment]
Can you gist your service configuration? I'll take a look at default interceptors to see if they cause any problems.

**Comment by paulvanbladel** [comment]
What is basically the simplest way to check if the interceptor is actually attached to $http ?

I'm trying to understand how this.options.bindRequestInterceptor is actually SET ?
Can you shed a light on this?
```
  // Setup request interceptors
    if (this.options.bindRequestInterceptor && isFunction(this.options.bindRequestInterceptor) &&
        this.options.bindResponseInterceptor && isFunction(this.options.bindResponseInterceptor)) {

      this.options.bindRequestInterceptor.call(this, this)
      this.options.bindResponseInterceptor.call(this, this)
    } else {
      // By default, request and response interceptors are for vue-resource
      this.$http.interceptors.push((request, next) => {
        if (this.isAuthenticated()) {
          request.headers.set('Authorization', [
            this.options.tokenType, this.getToken()
          ].join(' '))
        } else {
          request.headers.delete('Authorization')
        }
        
        next((response) => {
          try {
            var responseJson = JSON.parse(response[this.options.responseDataKey])
            if (responseJson[this.options.tokenName]) {
              this.setToken(responseJson)
              delete responseJson[this.options.tokenName]
              return responseJson
            }
          } catch(e) {}
          return response
        })
      })
    }
  }
```

**Comment by dgrubelic** [comment]
Let's switch to Gitter channel: https://gitter.im/vuejs-auth/vue-authenticate/
I'll be much easier to see what the problem is.

**Comment by dgrubelic** [comment]
Those interceptors are default ones setup for `vue-resource` library. That should work out of the box.

**Comment by paulvanbladel** [comment]
I'm trying to bring in my own interceptors, but that fails as well:
In below code: 
$auth.$http.interceptors.request is NULL !!!
So, interceptors exists, but request object is null.
```

...
Vue.use(VueResource)
Vue.use(VueAuthenticate, {
  bindRequestInterceptor ($auth) {
    debugger
    $auth.$http.interceptors.request.use((config) => {
      if ($auth.isAuthenticated()) {
        config.headers['Authorization'] = [
          $auth.options.tokenType, this.getToken()
        ].join(' ')
      }
      else {
        delete config.headers['Authorization']
      }
      return config
    })
  },

  bindResponseInterceptor ($auth) {
    $auth.$http.interceptors.response.use((response) => {
      $auth.setToken(response)
      return response
    })
  },
  baseUrl: 'http://localhost:8080',
  providers: {
    oauth2: {
      name: 'oauth2',
      url: 'Token/Exchange',
```

**Comment by vejja** [comment]
I am facing the same issue, basically the interceptors are not bound.
My config is as standard as it can be :

- bootstrap-vue
- vue-router
- vue-resource
- vue authenticate

The authentication part works perfectly. The token is returned and stored in localStorage. `isAuthenticated()` returns true. But the interceptors do not add the `Authorization : Bearer ...` header in the requests.

I looked quite extensively in the debugger and here are a couple takeways :
1. With the standard interceptors for vue-resource, the code gets into the section in lines 44-74 that start with the the comment `// Setup request interceptors`, then correctly goes into the `else {
      // By default, request and response interceptors are for vue-resource`. But the interceptors never get attached to `this.$http`

2. If tried using the configuration option `bindRequestInterceptor` with the example you give in the README. I changed the code and copied the one above (lines 44-74 of authenticate.js) so that it works with vue-resources. I have the same issue. I can tell the interceptor is not bound, because looking at `this` in the console, there is nothing under `options` for `bindRequestInterceptor`. It is just `null`.

3. The only way I would get it working was by using `Vue` directly with `Vue.http.interceptors.push()` and copying your code lines 44-74. in order to re-use across components I need to reimport Vue in each of these components, which feels wrong but works.

In a nutshell, when trying to bind interceptors to vue-resource, it looks like
- working with `this` doesn't work
- working with `Vue` works

Hope this helps

**Comment by vejja** [comment]
One side note, it could be completely unrelated but here is one strange thing I noticed in the `authenticate(provider)` method.

When the user is not authorized this method returns a reject promise, which is completely normal and you can catch the error in the `catch()` method. Great.

Now the weird thing is that within that catch, `this.$auth` is **undefined**. 

So :
```
this.$auth.authenticate('google')
.then(data => {
   // WITHIN THIS SCOPE THIS.$AUTH EXISTS OK
})
.catch(error => { 
  // BUT WITHIN THAT SCOPE THIS.$AUTH DOES NOT EXIST ??? 
})
```

**Comment by dgrubelic** [comment]
Any progress on this one? Have you managed to make it work?

BTW, i'm currently working on a replacing `vue-resource` with `axios` as a default request library (will be published in a few days). This will resolve these issues with interceptor bindings.

**Comment by demircancelebi** [comment]
Sadly, this is still broken. To make it work, you can use axios, vue-axios and custom interceptors together right now. As in:

```js
import axios from 'axios';
import VueAxios from 'vue-axios';
import VueAuthenticate from 'vue-authenticate';

Vue.use(VueAxios, axios);

Vue.use(VueAuthenticate, {
  bindRequestInterceptor() {
    this.$http.interceptors.request.use((config) => {
      if (this.isAuthenticated()) {
        config.headers.Authorization = [this.options.tokenType, this.getToken()].join(' ');
      } else {
        delete config.headers.Authorization;
      }
      return config;
    });
  },
  bindResponseInterceptor() {
    this.$http.interceptors.response.use((response) => {
      this.setToken(response);
      return response;
    });
  },
})
```

**Comment by dgrubelic** [comment]
Can you try now, I have replaced vue-resource as a default request handling library. Let me know if it works ok for you.

**Comment by dgrubelic** [comment]
Closing due to inactivity...

**Comment by Feng-Shui** [comment]
So I've run into this issue too, auth token is getting generated/saved but not attached to http requests in the Auth header.

Setting explicit `bindRequestInterceptor` and `bindResponseInterceptor` as per @demircancelebi didn't work for me, they were not being called similar to what @vejja was experiencing.

However, setting the token on axios did result in the request header being sent:

```
axios.defaults.headers.common['Authorization'] = 'Bearer [token value]' 
Vue.use(VueAxios, axios);

Vue.use(VueAuthenticate, {
    ...
  },
});
```

After a bit more digging around, I've discovered that if you call `$auth.isAuthenticated();` before any http requests are made, the token header is added:

```
const vm = new Vue({
  router,
  store,
  render: h => h(App),
}).$mount('#app');

vm.$auth.isAuthenticated();
```
